Adding DCP support for image based-transcriptomics (#70)
HumanCellAtlas/dcp-community: rfcs/text/0000-image-base-transcriptomics.md

# **Adding DCP support for image based-transcriptomics**

## **Summary**

The Human Cell Atlas (HCA) Governance Group (GG) has requested that the
Data Coordination Platform (DCP) prioritize the incorporation of an
image-based transcriptomics assay into the DCP. This RFC describes the
minimum set of deliverables that must be prioritized and completed for
the DCP to be able to (1) store and (2) process image-based single-cell
transcriptomics experiments. The following assumptions are made:

1.  Support for an image-based transcriptomics assay in the early
    iteration phase should be added to the DCP

2.  The current state of the starfish ecosystem is sufficient to

>     fulfill this need and will be selected by the AWG.

Reviewer: can you elaborate on what "this" is? Are you referring to #2 above?

Reviewer 2: ```suggestion
    process image-based transcriptomics experiments and will be selected by the AWG.
```

3.  This instantiation will require limited scale (low hundreds of
    datasets over several years)

## **Author(s)**

[Ambrose Carr](mailto:acarr@chanzuckerberg.com)

[Josh Moore](mailto:j.a.moore@dundee.ac.uk)

## **Shepherd**
[Brian Raymor](mailto:braymor@chanzuckerberg.com)

## **Motivation**

Cellular imaging assays typically either (1) categorize cells based on
the presence/absence of a particular set of markers, or (2) count
structures within cells. These assays exist at variable levels of
magnifications, including the counting of
[<span class="underline">individual
atoms</span>](https://www.sciencemag.org/news/2018/07/tour-de-force-researchers-image-entire-fly-brain-minute-detail)
(Electron Microscopy (EM))
[<span class="underline">synapses</span>](https://www.hhmi.org/news/how-to-rapidly-image-entire-brains-at-nanoscale-resolution)
(lattice light sheet), or [<span class="underline">counting molecules
within cells</span>](http://linnarssonlab.org/osmFISH/) (image-based
single-cell transcriptomics and image-based proteomics).

Image-based transcriptomics assays are a good candidate for DCP imaging

> support. They are relatively simple imaging assays that transform image

Reviewer 3: I do not believe you make a distinction between fluorescence microscopy and other spatial transcriptomics assays like ST and Seq-well. It may be important to point out the difference and that starfish would not support the latter (to my knowledge).

Reviewer 2: Those are sequencing based experiments, not image-based. Is that distinction too subtle?

Reviewer 3: It probably is the right distinction, but I am not sure it will be clear to all readers given informally it is all referred to as spatial and imaging. Use your best judgment here, the solution may be being more clear here or it may be working towards a broader awareness of the distinction.

features into a series of detected transcripts, which are aggregated by
cell. Because these assays require thin-sections of tissue, they produce
datasets that are relatively small by imaging standards (1-10s TBs). In
addition, image-based transcriptomics pipelines generate the same cell x
gene expression matrix outputted by scRNA-seq assays with additional
cellular annotations of the location of the cells. Because these data
structures are already present in the DCP, there are fewer components
that need to be generalized than for other imaging assays. Finally,
engineers from
[*<span class="underline">starfish</span>*](https://spacetx-starfish.readthedocs.io/en/latest/)
and
[<span class="underline">OME</span>](https://www.openmicroscopy.org/)
are working together to craft data formats, validation tooling, and
processing pipelines that comply with DCP requirements where possible.
These characteristics make this assay class a good choice for
prototyping image support in the DCP as it presents the smallest scale
challenge, the majority of the major deliverables can be provided by
DCP-extrinsic developers, and it requires generalization of the fewest

> DCP components. Implementation of support for these assays is likely to

Reviewer 3: The use of generalization is a little confusing. Later I know you get a little more into this but maybe simply saying it requires the least amount of updates or extension of DCP components is more direct?

generalize to additional areas of interest to the HCA community such as
multiplexed antibody imaging and thick-section volumetric imaging for

> both RNA and protein.

Reviewer 4: We have imaging based protein datasets ready now and CODEX planned from a few labs. We will need an explicit plan for protein. Imaging based transcriptomics is indeed the start of this support but can we add any specific deliverables related to this? Or add some timelines/plans that wranglers can pass on to contributors?

Reviewer 2: Starfish's roadmap does not include proteomics assays at this time. Our _format_ supports the storage of those data, and starfish would welcome the contribution of a pipeline. However, proteomics is not in-scope for this RFC. 

> 

> We have imaging based protein datasets ready now [...] can we add any specific deliverables related to this? Or add some timelines/plans that wranglers can pass on to contributors?

Could you start an email thread about image-based proteomics? I don't want to expand the scope of this RFC, but agree this is an important question, am interested in what data you have, and want to consider soliciting one or more pipelines to process those data.

    initiate imaging support in the DCP because of their similarity to
    existing DCP sequencing assays and the abundance of labs using
    single-molecule FISH assays to clarify spatial distributions of
    cell types identified in RNA-seq experiments.

  - If desired, *all* key validation, specification, and analysis
    deliverables can be provided by *starfish* and OME engineers,
    minimizing load on DCP engineering teams.

  - A draft metadata schema has already been built by EBI via
    collaboration with the SpaceTx consortium, who have delivered
    image-based transcriptomics datasets which the ingestion service
    is beginning to test.

  - Based on [<span class="underline">delivery timelines for
    prerequisite
    work</span>](https://docs.google.com/document/d/1RQcj450v0WHGUvSm7YwwFQ7sRFoPAOCjRbo63UzyQd0/edit?ts=5c40d3f1#heading=h.4qbf4q96id16),
    implementation of support for image-based transcriptomics could
    begin as early as Q2-Q3 2019.

  - While image-based transcriptomics data can be ingested and stored
    without a visualization capability, significant user demand can be
    expected similar to that for count matrices.

### **User Stories**

>   - As a submitter, I want to upload directories of files representing

Reviewer 2: ```suggestion
  - As a submitter, I want to upload directories of files in SpaceTx Format representing
```

>     an imaging experiment for storage in the DCP.

Reviewer 4: This should explicitly include conversion from local formats into starfish formats.

>     an imaging experiment for storage in the DCP.

Reviewer 2: ```suggestion
    an imaging experiment for storage in the DCP so that I can participate in the glorious endeavor of building an atlas of human cells
```

  - As a user with a keyboard, I want to identify images with a given

>     gene for download.

Reviewer 2: ```suggestion
    gene for download so that I can visualize the gene's spatial patterning.
```

>   - As a user with a keyboard, I want to see spots overlayed on an

Reviewer 2: ```suggestion
  - As a user with a keyboard or pipette, I want to see spots overlayed on an
```

>     image for QC.

Reviewer 3: I would argue this is also a user with a pipette.

>     image for QC.

Reviewer 2: ```suggestion
    image for QC to convince myself that the data have been properly processed.
```

  - As a user with a keyboard, I want to see cells overlayed on an

>     image for QC.

Reviewer 2: ```suggestion
    image for QC to confirm that cells have been properly segmented.
```

  - As a user with a keyboard, I want to generate a gene x cell matrix
    for biological analysis of my data and combination with sequencing

>     assays.

Reviewer 2: ```suggestion
    assays so that I can use these data to answer biological questions.
```

## **Scientific "guardrails"**

Image-based transcriptomics assays are still in active development, and
benchmarking and characterization of these data types is in process. As

> such we expect to utilize community feedback to iterate on scientific

Reviewer 5: Who is the "we" here? Does that mean Ambrose and Josh?

Author: Thanks, @mckinsel. @ambrosejcarr slipped one past me in a late revision! I'll let him disambiguate.

> such we expect to utilize community feedback to iterate on scientific

Reviewer 2: ```suggestion
such starfish and the Data Processing Pipelines team expect to utilize community feedback to iterate on scientific
```
@kbergin @TimothyTickle Are you being OK with this language and attribution?

proposed integration should be reviewed against the HCA Science
governance’s understanding of both the communities likely use of imaging
data as well as the long-term costs of storage and transfer. Based on
the intended investment for imaging, additional effort at this initial
stage of the architecture may be called for to increase value or
decrease cost.

## **Detailed Design**

### Data Specification

A data specification consists of format specifications for any files
(inputs and outputs) used by imaging assays, and HCA metadata that sits
on top of these data and describes information about the samples and how
the data were acquired. Formats and specifications listed below are
being developed by *starfish*, OME, and HCA Metadata teams. Each have
working prototypes supported by active development teams.

#### Input Data Formats

##### SpaceTx Format

To prototype the use of imaging data in the DCP, the *starfish* and OME
teams propose that the DCP use [<span class="underline">SpaceTx
Format</span>](https://spacetx-starfish.readthedocs.io/en/latest/sptx-format/index.html),
a format that groups and indexes a potentially very large number of two

> dimensional images, forming 5-dimensional tensors over (time, channel,

Reviewer: I am not sure what a "channel" is here. Should that be defined? Is it a color channel?

> dimensional images, forming 5-dimensional tensors over (time, channel,

Reviewer 2: ```suggestion
dimensional images, forming 5-dimensional tensors over (time, fluorescence channel,
```

z, y, x). These dimensions are fairly consistent across imaging domains
and will likely generalize to other use cases. SpaceTx created this
format because they were unable to find an existing format that
satisfied our cloud compute and data visualization use cases. *starfish*
and OME have since become aware of contemporaneous development of
projects that aim to satisfy similar use cases and are working towards
unification, where possible across scientific disciplines and languages.

To avoid writing a large amount of microscope-specific code at the
outset, *starfish* requests that users carry out basic pre-processing of
their data to fix registration or chromatic errors. More details on the
specific requests made of users can be found
[<span class="underline">here</span>](https://drive.google.com/a/chanzuckerberg.com/open?id=1x516WMMWkEKKQwLrqqBsKmsPEDNjFYugqxm400D2NjY).
The exact specification of the image data format will depend upon the

> delivery date that the DCP requests for this prototype.

Reviewer 3: How do we recover provenance for the microscope-specific code and are there avenues for standardizing (even if stratified by microscope or company)?

Reviewer 2: We're working to include these microscope-specific steps for the reasons you outline, and because we want to make sure to capture "actually-raw" data. 

Some of these steps can be standardized by requiring specific additional input data from users (e.g. a picture of the point spread function of each channel of a camera would enable standardized deconvolution). However, we need to do additional research to understand what kinds of data our data generators will be willing to provide. 

The best chance at standardization will come from commercialization, through which data is generated by a single device.

Reviewer 3: For now, let's just make sure to describe where the user started and what they did thereafter which sounds like what you are focused on.

##### Pipeline Recipes

The parameterization of an imaging pipeline depends on the assay in
question, the microscope used to capture the data, the density of spots
in the tissue, and the tissue being analyzed. For example, if the same
assay is run a second time at twice the magnification, spots would be
expected to have twice the radius, which would need to be communicated
to the spot-finding algorithm.

Human analysts are very good at selecting these parameters, but
processes to fit parameters to all datasets across assays, tissues, and
microscopes have not yet been developed. To solve this problem, starfish
exposes the ability for users to pilot starfish on small subsets of
their data in a local environment. These analyses generate parameter
sets ("pipeline recipes") which are stored in json and validated by
*starfish*. These parameters are then used to run *starfish* across the
complete dataset, in a cloud-based environment.

#### Output Data Formats

Image-based transcriptomics experiments emit three output formats that
contain identified features, all of which can be emitted in either

> netcdf or zarr file formats, the latter being the format consumed by the

Reviewer 3: Is netcdf compatible with object-store? (We have confirmed zarr is)

Reviewer 2: It is a single file; I am not aware of anything that would make it incompatible.

Reviewer 3: great

Image-based transcriptomics experiments image the same tissue using
multiple color channels, and after subjecting the tissue to different
reagents. The physical locations of fluorescent molecules manifest as
spots in the images, and these spots are detected and aggregated across
fluorescence channels and time to build codes. In practice, the process
of detecting and decoding spots contains significant possibilities for
error, and the IntensityTable supports analysis and QC for this
process.

##### [SegmentationMask]

In order to assign the spots identified above to individual cells, the
locations of cells within the images must be identified. Compressed
segmentation masks are a compact way to represent these data, which
store, for each cell, the portion of the image that is attributed to
that cell.

##### [ExpressionMatrix]

*starfish* combines spot calls from the IntensityTable with cell areas
in the SegmentationMask to produce expression matrices. These are saved
in zarr format and look identical to the matrices emitted by the Data
Analysis Service pipelines, with the exception that each cell is
additionally annotated with metadata that specify the x, y, and z

> location of the data.

Reviewer 3: Will the names of the genes/features in the expression matrix be required to be consistent with naming from the standard transcriptome reference of the DCP?

Reviewer 2: Yes.

> location of the data.

Reviewer 2: ```suggestion
location of the data. Feature names will be consistent with DCP standards.
```

##### Processed Images

Image-based transcriptomics experiments suffer from several sources of
error and variation which are removed via image processing. Microscope
stages or tissue slices may shift as reagents are added and removed
across imaging rounds, or which spectral filters are swapped between
channels. In addition, reagents can non-specifically bind to cellular
material, causing autofluorescence. Prior to spot-calling, images can be
aligned, background removed, and spots computationally enhanced.
*starfish* can output these processed images, annotated with information
sufficient to re-do the processing, so that the images input to spot
calling can be evaluated by users. If emitted, processed images will be
generated in SpaceTx Format.

#### HCA Sample Metadata

The HCA metadata team has built a metadata specification draft for
imaging data that references data in SpaceTx-Format. The issue that
formed the draft can be found

> [<span class="underline">here</span>](https://github.com/HumanCellAtlas/metadata-schema/issues/368).

Reviewer 4: Requirements here included:
- Generality for experiments including immunohistochemistry, in situ sequencing and other FISH methods.
- Multiplexed and non-multiplexed experiments.
- Target to probe mapping
- Target level metadata (search by gene)
- Congruence with current entity model
- Explicit link to experiment.json which links to all other data files.
- One experiment.json per assay process (aka bundle)

Reviewer 2: > * One experiment.json per assay process (aka bundle)

Is "assay process" consistent with a "replicate" as we think of in sequencing? There may be multiple `experiment.json` files per study. 

Everything else looks good. 👍

Reviewer 6: @ambrosejcarr I suggest the add a layer into experiment.json such as run.json (or some other name) to maintain the one-to-one reference of experiment - project(study).

Reviewer 2: This sounds like a good feature request for starfish. Making sure I understand, the key objects the HCA needs are: 

1. `experiment.json` (top-level project)
2. `run.json` (a complete collage of a tissue sample containing many fields of views and potentially multiple imaging modalities (e.g. DIC))
3. `field_of_view.json` (a piece of the collage)

Our manifest.json could probably be repurposed to store the `run.json` If that's right I'll make an issue and reference it here.

Reviewer 4: Zina's suggestion is fine with me, though there are multiple solutions. At the interface between spacetx metadata and hca metadata the main requirement is good definitions on entities that contributors and wranglers can abide by. 'experiment.json' is too loosely defined at the moment (maybe by design) to be relied upon to be consistent across datasets. In this informational RFC I think it is worth adding some detail to this section as I listed above and ensuring that our definition of experiment is clearly defined.

Reviewer 4: @ambrosejcarr I suggest adding: 
1. Metadata design requirements as bulleted above (maybe these could be written in a clearer more accessible way)
2. SpaceTx entity definitions of experiment, FOV etc. as these affect how hca metadata is linked to the data. Changing these definitions would need us to revisit how we link data to metadata.

### Data and Metadata validation

#### Input Data validation

SpaceTx Format consists of a json file that specifies how to compose a
dataset from a set of two-dimensional images. As a result, it is
possible to rapidly validate that a file is complete and internally
consistent. In addition, SpaceTx Format stores per-chunk checksum
information, enabling the data to be checked for integrity. It is
possible to carry out additional, more data-specific checks, such as
that images are not blank, contain spots, or contain cells, however
these checks take increasing amounts of time.

OME and SpaceTx can provide a docker container that validates image
datasets with the highest level of stringency possible in a DCP-provided
"maximum validation time".

#### Output Data validation

Each of the output formats produced by SpaceTx can be stored in Zarr and
read (on-demand) by *starfish* and xarray. SpaceTx provides a docker
container that loads and validates each output file using the xarray
library. Validation can be tuned to occur within a DCP-provided target

> validation time.

Reviewer 3: What do you mean by "can be tuned to occur within a ...time"?

Reviewer 2: If we told "validation can cost $1 or take 30s" starfish will validate the json files. If it can cost many dollars and take hours, we can validate the image data. 

If something in between, we could randomly test a few images and all the json files.

Reviewer 2: If we told "validation can cost $1 or take 30s" starfish will validate the json files. If it can cost many dollars and take hours, we can validate the image data. 

If something in between, we could randomly test a few images and all the json files.

Reviewer 3: ok, thanks for the clarification.

### Data Processing Workflows

To process data in the DCP, the *starfish* team can deliver WDL-wrapped
data processing workflows for any of the existing image-based
transcriptomics pipelines. These workflows are based on the *starfish*
python library, which in turn leverages standard open-source python
projects. *starfish* provides a docker container that exposes both a
python API and a CLI, and supports composable, modular, pipelines. These
pipelines take as inputs images in SpaceTx Format and a pipeline recipe.
Each assay emits output in the formats specified above. More detail on
the state of image-based transcriptomics analyses can be found
[<span class="underline">here</span>](https://docs.google.com/document/d/18Ve9C_cEn5lI2paivfgTAyjsM62qMyUF_s4mj82l1L0/edit#).

### Data Processing Infrastructure

> All *starfish* pipelines share the same inputs and outputs. It should

Reviewer 3: Are these multiple workflows or one workflow parameterized for multiple assays (preferred)?

Reviewer 2: At this time it makes more sense to deliver multiple wdls. This may be reduced to a single wdl in the future. 

Resolving this was unintentional.

therefore be possible to deliver a single adapter WDL that satisfies any
SpaceTx assay, and a family of subscriptions with minimal inter-class
changes that specify the particular assay in question.

### Delivery of Output Data to Users

#### Matrices

As mentioned above, image-based transcriptomics output data adhere to
existing Matrix Service specifications and can be served "as is" for the
imaging prototype. Additional information on cellular geometries and QC
will be provided as a supplementary metadata zarr archive.

However, other imaging experiments will likely require a generalization

> of the internal namespace of the DCP. Currently, the DCP standardizes

Reviewer 3: Using "internal namespace of the DCP" makes me feel like you are talking about domain spaces and is confusing. Maybe "will likely requiring extending how features (genes) are named to a support a broader set of features (eg. proteins and cellular components).

Reviewer 2: @josh can you include Tim's wording in the next pass?

around GENCODE gene IDs and HGNC gene symbols. Image based proteomics
will require protein identifiers such as UNIPROT, and other use cases
that measure sub-cellular components will likely require more general
cellular component information (e.g. "endoplasmic reticulum"). These use
cases could be supported by switching identifiers from, e.g. "ACTB" to
"HGNC:ACTB" by prepending the namespace of the IDs being used in a given

> assay.

Reviewer 3: I agree with the general idea but may not agree with the specifics here, either way, please remove the implementation detail. The time will have time in the future to explore this and other solutions. (from "These use" ... to ... "assay".)

### **Acceptance Criteria**

Except where listed ("responsible team"), the starfish and OME teams
will do the majority of project management and execution for the
following deliverables. They will coordinate and collaborate with DCP
partners ("contact person") to validate design correctness and minimize
code changes and their impact.

> | Component/Deliverable                                 	| Responsible Team/Service 	| No earlier than 	| Contact Person	|

Reviewer 3: Should this table be complete before approval?

Reviewer 2: I think the contacts should be completed. The "not delivered before" was meant to help the DCP teams understand the readiness that we're aware of, but I think the actual delivery dates are subject to the prioritization of the work in this RFC, so I'm not too concerned about having those filled in.

| Submission Datasets Available                         	| starfish                 	| Q2 2019         	| A. Carr       	|
| Analysis Pipelines                                    	| Broad                    	| Q2 2019         	| ?             	|
| Analysis & Adapter Pipelines                          	| Broad                    	| Q2 2019         	| A. Carr       	|
| Lira subscription                                     	| Broad                    	| TBD             	| S. Ehsan      	|
| Data Analysis Service starfish Pipeline Tests         	| Broad                    	| Q2 2019         	| G. Grant      	|
| Specification for input format                        	| starfish                 	| Complete        	| J. Moore      	|
| Specification for output format                       	| starfish                 	| Q1 2019         	| A. Carr       	|
| Input validation tool                                 	| starfish, OME            	| Q3 2019         	| J. Moore      	|
| Output validation tool                                	| starfish, OME            	| Q2 2019         	| A. Carr       	|

> | Ingest support for the nested arrays of channels      	| EBI                      	| ?               	| ?             	|

Reviewer 4: Zina can be added here too

Reviewer 2: ```suggestion
| Ingest support for the nested arrays of channels      	| EBI                      	| ?               	| Z. Perova             	|
```

| [Upload and download of subdirectories][b]             	| ?                        	| ?               	| ?             	|
| DSS spatial data bundle (or bundless) specification   	| UCSC                     	| ?               	| B. Hannafious 	|
| DSS scale testing                                     	| UCSC / DCP               	| TBD             	| B. Hannafious 	|
| Data Browser facets for spatial data                  	| UCSC                     	| TBD             	| H. Schmidt    	|
| Matrix service integration                            	| starfish                 	| TBD             	| M. Kinsella   	|
| Upload Bundle for 1M+ files                           	| ?                        	| ?               	| P. Shah       	|
| Image file download latency testing                   	| ?                        	| ?               	| ?             	|

### **Unresolved Questions**

  - Who are the intended beta users for imaging support and what
    functionality is being targeted?

  - Are there other roadblocks for imaging support in the DCP?

  - What, if any, additional automation may be needed in the DCP to

>     support imaging?

Reviewer 4: Support (tech and wrangling) for contributors outside of the SpaceTx project to convert to this format or support for ingestion of other formats.

Reviewer 2: @joshmoore can you add a brief section that covers our intentions with respect to Bio-Formats?

Author: @hewgreen : let me know what you think about 6d3eabfbb0b9f06a97d405a6bdb2ced00b3b589c -- i.e. added as a drawback with existing efforts for remediation.

Reviewer 4: Thanks Josh, looks good. At this stage predicting this requirement is sufficient because it isn't clear if this will be frequently required, prove too difficult or produce an interoperable result.

  - What is the appropriate separation of metadata in the SpaceTx

>     Format and the HCA metadata schema?

Reviewer 4: As a high level description, HCA contains metadata about biological entities and protocols used to make the imaged specimen. HCA also has very limited file metadata (name, format, type). HCA will also contain target and probe information to support search. This is may be repeated in the SpaceTx codebook and probes.sam which introduces a validation requirement to ensure continuity. SpaceTx metadata contains more granular fields related to the image. These fields are typically also captured by OME formats. Critically metadata is linked via SpaceTx's experiment.json explicitly (imaged_specimen -> experiment.json -> the rest of the data).

Reviewer 2: Ok. Sounds to me like this question can safely be removed. Thanks Matt.

  - How should imaging metadata be validated?

### **Drawbacks and Limitations**

Imaging formats and pipelines have not yet reached a level of
standardization across the community. As such, flexibility around
updating and upgrading formats and pipelines will be necessary.
Specifically:

1.  The development of a chunked image file format\[1\] that supports
    cloud workflows for data at the scale that will become common over
    the next 10 years is in an early stage. Working prototypes from
    genomics, imaging, astronomy, and geosciences exist and a process
    is beginning to unify formats. However, this process will evolve
    over time, and it will be necessary to upgrade the data,
    potentially multiple times, as backwards-incompatible changes are
    made. To mitigate this risk, the OME and *starfish* teams are
    willing to be responsible for producing containerized scripts to
    upgrade data when this is necessary, and to adhere to schedules
    that limit the frequencies of these updates.

> 

Reviewer: The data store team will need to know more about the data structures you would like to use. (Note that I was unable to attend the imaging breakout at the last F2F, but I don't see a suggested first design here). So far we are working on supporting millions of files per bundle, but is that what you need?

Reviewer 2: Does this documentation answer your questions?[https://spacetx-starfish.readthedocs.io/en/stable/sptx-format/index.html](https://spacetx-starfish.readthedocs.io/en/stable/sptx-format/index.html)

If yes, would you like to see this reference included?

Reviewer 2: I've updated the link in the reference section `#### Input Data Formats` to contain our design information on SpaceTx-Format. I'll resolve this, but please re-open if you'd like more information added.

matrix service. In addition, it has the opportunity to emit image files
that have been processed to align images, remove autofluorescence and
enhance spots. All output formats (including SpaceTx image format)
contain full provenance records of the analysis environment, versions of
starfish and key dependencies, algorithms that were run, and their
parameters.

##### [IntensityTable]

Image-based transcriptomics experiments image the same tissue using
multiple color channels, and after subjecting the tissue to different
reagents. The physical locations of fluorescent molecules manifest as
spots in the images, and these spots are detected and aggregated across
fluorescence channels and time to build codes. In practice, the process
of detecting and decoding spots contains significant possibilities for
error, and the IntensityTable supports analysis and QC for this
process.

##### [SegmentationMask]

In order to assign the spots identified above to individual cells, the
locations of cells within the images must be identified. Compressed
segmentation masks are a compact way to represent these data, which
store, for each cell, the portion of the image that is attributed to

> that

Reviewer 7: Remove line break between "that" and "cell".

on top of these data and describes information about the samples and how
the data were acquired. Formats and specifications listed below are
being developed by *starfish*, OME, and HCA Metadata teams. Each have
working prototypes supported by active development teams.

#### Input Data Formats

##### SpaceTx Format

To prototype the use of imaging data in the DCP, the *starfish* and OME
teams propose that the DCP use [<span class="underline">SpaceTx
Format</span>](https://spacetx-starfish.readthedocs.io/en/latest/sptx-format/index.html),
a format that groups and indexes a potentially very large number of two
dimensional images, forming 5-dimensional tensors over (time, fluorescence channel,
z, y, x). These dimensions are fairly consistent across imaging domains
and will likely generalize to other use cases. SpaceTx created this
format because they were unable to find an existing format that
satisfied our cloud compute and data visualization use cases. *starfish*
and OME have since become aware of contemporaneous development of
projects that aim to satisfy similar use cases and are working towards
unification, where possible across scientific disciplines and languages.

> To avoid writing a large amount of microscope-specific code at the

Reviewer 6: could I please get access to the mentioned doc?

Reviewer 2: I shared the original doc with the wranglers. We've also uploaded an abridged version that removes SpaceTx-consortium specific information here that has all the relevant information: https://spacetx-starfish.readthedocs.io/en/latest/usage/index.html

I have added it to the RFC.

outset, *starfish* requests that users carry out basic pre-processing of
their data to fix registration or chromatic errors. More details on the
specific requests made of users can be found

> [<span class="underline">here</span>](https://drive.google.com/a/chanzuckerberg.com/open?id=1x516WMMWkEKKQwLrqqBsKmsPEDNjFYugqxm400D2NjY).

Reviewer 2: ```suggestion
[<span class="underline">here</span>](https://spacetx-starfish.readthedocs.io/en/latest/usage/index.html).
```

starfish and key dependencies, algorithms that were run, and their
parameters.

##### [IntensityTable]

Image-based transcriptomics experiments image the same tissue using
multiple color channels, and after subjecting the tissue to different
reagents. The physical locations of fluorescent molecules manifest as
spots in the images, and these spots are detected and aggregated across
fluorescence channels and time to build codes. In practice, the process
of detecting and decoding spots contains significant possibilities for
error, and the IntensityTable supports analysis and QC for this
process.

##### [SegmentationMask]

In order to assign the spots identified above to individual cells, the
locations of cells within the images must be identified. Compressed
segmentation masks are a compact way to represent these data, which
store, for each cell, the portion of the image that is attributed to
that cell.

##### [ExpressionMatrix]

*starfish* combines spot calls from the IntensityTable with cell areas
in the SegmentationMask to produce expression matrices. These are saved
in zarr format and look identical to the matrices emitted by the Data

> Analysis Service pipelines, with the exception that each cell is

Reviewer 3: Change Data Analysis Service pipelines to HCA data processing pipelines (as per charter title)

| Analysis Pipelines                                    	| Broad                    	| Q2 2019         	| ?             	|
| Analysis & Adapter Pipelines                          	| Broad                    	| Q2 2019         	| A. Carr       	|
| Lira subscription                                     	| Broad                    	| TBD             	| S. Ehsan      	|
| Data Analysis Service starfish Pipeline Tests         	| Broad                    	| Q2 2019         	| G. Grant      	|
| Specification for input format                        	| starfish                 	| Complete        	| J. Moore      	|
| Specification for output format                       	| starfish                 	| Q1 2019         	| A. Carr       	|
| Input validation tool                                 	| starfish, OME            	| Q3 2019         	| J. Moore      	|
| Output validation tool                                	| starfish, OME            	| Q2 2019         	| A. Carr       	|
| Ingest support for the nested arrays of channels      	| EBI                      	| ?               	| Z. Perova             	|

> | [Ability to describe files that belong together][a]   	| ?                        	| ?               	| ?             	|

Reviewer 2: ```suggestion
| [Ability to describe files that belong together][a]   	| ?                        	| ?               	| H. Schmidt             	|
```

Reviewer 2: @hannes-ucsc if you're not the right contact for this, could you suggest someone else?

> | [Upload and download of subdirectories][b]             	| ?                        	| ?               	| ?             	|

Reviewer 2: ```suggestion
| [Upload and download of subdirectories][b]             	| ?                        	| ?               	| H. Schmidt             	|

Reviewer 2: @hannes-ucsc if you're not the right contact for this, could you suggest someone else?

| DSS spatial data bundle (or bundless) specification   	| UCSC                     	| ?               	| B. Hannafious 	|
| DSS scale testing                                     	| UCSC / DCP               	| TBD             	| B. Hannafious 	|
| Data Browser facets for spatial data                  	| UCSC                     	| TBD             	| H. Schmidt    	|
| Matrix service integration                            	| starfish                 	| TBD             	| M. Kinsella   	|
| Upload Bundle for 1M+ files                           	| ?                        	| ?               	| P. Shah       	|

> | Image file download latency testing                   	| ?                        	| ?               	| ?             	|

Reviewer 2: ```suggestion
| Image file download latency testing                   	| ?                        	| ?               	| B. Hannafious             	|

Reviewer 2: @xbrianh if you're not the right contact for this, could you suggest someone else?

### **Unresolved Questions**

  - Who are the intended beta users for imaging support and what
    functionality is being targeted?

>   - Are there other roadblocks for imaging support in the DCP?

Reviewer 6: We need to have a clearly defined process for wrangling image-based transcriptomics datasets.

Reviewer 2: Good point. What's the right next step for this? An issue in a repository owned by the wranglers? Where have you defined the process for wrangling sequencing data?

Reviewer 6: @ambrosejcarr we have documentation in the wrangling repo on the process of wrangling datasets in general. I also shared this [Google doc](https://docs.google.com/document/d/1pc2h7pv5ayDm8VOxu2aN3OrEtQtXi_v6ifZKDxp4-_Y/edit) adapted to wrangling image-based transcriptomics datasets that require SpaceTx formatting. I will be very happy to hear feedback from the involved/mentioned parties.

>   - Are there other roadblocks for imaging support in the DCP?

Reviewer 2: ```suggestion
  - Clearly define a wrangling process for imaging data.
```

transcriptomics pipelines. These workflows are based on the *starfish*
python library, which in turn leverages standard open-source python
projects. *starfish* provides a docker container that exposes both a
python API and a CLI, and supports composable, modular, pipelines. These
pipelines take as inputs images in SpaceTx Format and a pipeline recipe.
Each assay emits output in the formats specified above. More detail on
the state of image-based transcriptomics analyses can be found
[<span class="underline">here</span>](https://docs.google.com/document/d/18Ve9C_cEn5lI2paivfgTAyjsM62qMyUF_s4mj82l1L0/edit#).

### Data Processing Infrastructure

All *starfish* pipelines share the same inputs and outputs. It should
therefore be possible to deliver a single adapter WDL that satisfies any
SpaceTx assay, and a family of subscriptions with minimal inter-class
changes that specify the particular assay in question.

### Delivery of Output Data to Users

#### Matrices

As mentioned above, image-based transcriptomics output data
include a gene expression matrix and related metadata that are
compatible with matrix service requirements and should be no

> more difficult to convert than the output of a sequencing assay

Reviewer 8: NIt - missing "."

However, other imaging experiments will likely require extending how
features (genes) are named to support a broader set of features
(eg. proteins and cellular components). Currently, the DCP standardizes
around GENCODE gene IDs and HGNC gene symbols. Image based proteomics
will require protein identifiers such as UNIPROT, and other use cases
that measure sub-cellular components will likely require more general
cellular component information (e.g. "endoplasmic reticulum").

### **Acceptance Criteria**

Except where listed ("responsible team"), the starfish and OME teams
will do the majority of project management and execution for the
following deliverables. They will coordinate and collaborate with DCP
partners ("contact person") to validate design correctness and minimize
code changes and their impact.

> | Component/Deliverable                                 	| Responsible Team/Service 	| No earlier than 	| Contact Person	|

Reviewer 8: May I suggest that you eliminate the _No earlier than_ column and treat this table as required deliverables that will get scheduled (and perhaps get their own technical RFC) based on roadmap priorities? Per the RFC process - *NOTE*: _The existence of an approved software RFC does not indicate any particular priority or commitment to implement the RFC, nor is an approved RFC a green light to implement. Approved RFCs simply demonstrate community agreement on a technical decision. Product priorities are managed and implementations scheduled via the DCP PM planning process_ - and an explanation about "?" would be helpful.

Reviewer 2: Done.

| Output validation tool                                	| starfish, OME            	| Q2 2019         	| A. Carr       	|
| Ingest support for the nested arrays of channels      	| EBI                      	| ?               	| Z. Perova           |
| [Upload and download of subdirectories][b]             	| ?                        	| ?               	| H. Schmidt           |
| DSS spatial data bundle (or bundless) specification   	| UCSC                     	| ?               	| B. Hannafious 	|
| DSS scale testing                                     	| UCSC / DCP               	| ?             	| B. Hannafious 	|
| Data Browser facets for spatial data                  	| UCSC                     	| ?             	| H. Schmidt    	|
| Matrix service integration                            	| starfish                 	| ?             	| M. Kinsella   	|
| Upload Bundle for 1M+ files                           	| ?                        	| ?               	| P. Shah       	|
| Image file download latency testing                   	| ?                        	| ?               	| B. Hannafious             	|

### **Unresolved Questions**

>   - Who are the intended beta users for imaging support and what

Reviewer 8: How do you plan to resolve this question?

Reviewer 2: I removed this question, but the short answer is that it's implicit in the outlined features: The lack of visualization support means we're targeting users who are comfortable interacting with count matrices or raw image data. These is our _user with a keyboard_.

    functionality is being targeted?

>   - Clearly define a wrangling process for imaging data.

Reviewer 8: Seems like wrangling process could be added to the table above?

Reviewer 6: @brianraymor @ambrosejcarr I suggest adding "Transformation of data into the accepted format" rather than wrangling process itself. We have a defined wrangling process, however, the imaging data requires transformation into the SpaceTx format in order to be wrangled. The wranglers do not have the capacity to do it, and last week's SpaceTx Hackathon showed that it is highly dependent on the data and requires a good understanding of both the pipeline and the experimental approach. Therefore I think it is crucial to determine who will be responsible for this. I see two options: to find a more simple way to transform the data and expect data contributors to do it themselves, or allocate resources for a dedicated person to do it for all the data coming into HCA.

Reviewer 2: Added.

>   - What, if any, additional automation may be needed in the DCP to

Reviewer 8: Recommend cut. Reviewers should identify in your deliverables table as needed.

Reviewer 2: Removed.

    support imaging?

>   - How should imaging metadata be validated?

Reviewer 8: Again, how do you plan to resolve this question or would it be a future RFC?

Reviewer 2: Future RFC, removed.

# **Adding DCP support for image based-transcriptomics**

## **Summary**

The GG has requested that the DCP prioritize the incorporation of an
image-based transcriptomics assay into the DCP. This RFC describes the
minimum set of deliverables that must be prioritized and completed for
the DCP to be able to (1) store and (2) process image-based single-cell
transcriptomics experiments. The following assumptions are made:

1.  Support for an image-based transcriptomics assay in the early
    iteration phase should be added to the DCP

2.  The current state of the starfish ecosystem is sufficient to
    fulfill this need and will be selected by the AWG.

3.  This instantiation will require limited scale (low hundreds of
    datasets over several years)

## **Author(s)**

[Ambrose Carr](mailto:acarr@chanzuckerberg.com)

[Josh Moore](mailto:j.a.moore@dundee.ac.uk)

## **Shepherd**

> 

Reviewer 2: ```suggestion
[Brian Raymor](mailto:braymor@chanzuckerberg.com)

```

    initiate imaging support in the DCP because of their similarity to
    existing DCP sequencing assays and the abundance of labs using
    single-molecule FISH assays to clarify spatial distributions of
    cell types identified in RNA-seq experiments.

  - If desired, *all* key validation, specification, and analysis
    deliverables can be provided by *starfish* and OME engineers,
    minimizing load on DCP engineering teams.

  - A draft metadata schema has already been built by EBI via
    collaboration with the SpaceTx consortium, who have delivered
    image-based transcriptomics datasets which the ingestion service
    is beginning to test.

  - Based on [<span class="underline">delivery timelines for
    pre-requisite
    work</span>](https://docs.google.com/document/d/1RQcj450v0WHGUvSm7YwwFQ7sRFoPAOCjRbo63UzyQd0/edit?ts=5c40d3f1#heading=h.4qbf4q96id16),
    implementation of support for image-based transcriptomics could
    begin as early as Q2-Q3 2019.

  - While image-based transcriptomics data can be ingested and stored
    without a visualization capability, significant user demand can be
    expected similar to that for count matrices.

> ### **User Stories**

Reviewer 9: These user stories are missing the `why/benefit/goal` section. Please add them.

Reviewer 9: Here's a link for reference:

https://en.wikipedia.org/wiki/User_story#Common_templates

to the spot-finding algorithm.

Human analysts are very good at selecting these parameters, but
processes to fit parameters to all datasets across assays, tissues, and
microscopes have not yet been developed. To solve this problem, starfish
exposes the ability for users to pilot starfish on small subsets of
their data in a local environment. These analyses generate parameter
sets ("pipeline recipes") which are stored in json and validated by
*starfish*. These parameters are then used to run *starfish* across the
complete dataset, in a cloud-based environment.

#### Output Data Formats

Image-based transcriptomics experiments emit three output formats that
contain identified features, all of which can be emitted in either
netcdf or zarr file formats, the latter being the format consumed by the
matrix service. In addition, it has the opportunity to emit image files
that have been processed to align images, remove autofluorescence and
enhance spots. All output formats (including SpaceTx image format)
contain full provenance records of the analysis environment, versions of
starfish and key dependencies, algorithms that were run, and their
parameters.

> ##### [<span class="underline">IntensityTable</span>](https://spacetx-starfish.readthedocs.io/en/ajc-output-specifications/sptx-format/output_formats/IntensityTable/index.html)

Reviewer 9: These title links are broken(?)

Author: Thanks, @mweiden. That branch got merged. Links updated.

validation time.

### Data Processing Workflows

To process data in the DCP, the *starfish* team can deliver WDL-wrapped
data processing workflows for any of the existing image-based
transcriptomics pipelines. These workflows are based on the *starfish*
python library, which in turn leverages standard open-source python
projects. *starfish* provides a docker container that exposes both a
python API and a CLI, and supports composable, modular, pipelines. These
pipelines take as inputs images in SpaceTx Format and a pipeline recipe.
Each assay emits output in the formats specified above. More detail on
the state of image-based transcriptomics analyses can be found
[<span class="underline">here</span>](https://docs.google.com/document/d/18Ve9C_cEn5lI2paivfgTAyjsM62qMyUF_s4mj82l1L0/edit#).

### Data Processing Infrastructure

All *starfish* pipelines share the same inputs and outputs. It should
therefore be possible to deliver a single adapter WDL that satisfies any
SpaceTx assay, and a family of subscriptions with minimal inter-class
changes that specify the particular assay in question.

### Delivery of Output Data to Users

#### Matrices

> As mentioned above, image-based transcriptomics output data adhere to

Reviewer 10: This may not be true anymore. @mckinsel is best positioned to provide additional details, but the matrix service new implementation no longer reads in zarr stores from dss bundles.

Reviewer 2: Makes sense. @joshmoore I think it's probably sufficient to replace this paragraph with:

```
As mentioned above, image-based transcriptomics output data 
include a gene expression matrix and related metadata that are 
compatible with matrix service requirements and should be no
more difficult to convert than the output of a sequencing assay
```

"HGNC:ACTB" by prepending the namespace of the IDs being used in a given
assay.

### **Acceptance Criteria**

Except where listed ("responsible team"), the starfish and OME teams
will do the majority of project management and execution for the
following deliverables. They will coordinate and collaborate with DCP
partners ("contact person") to validate design correctness and minimize
code changes and their impact.

<table>
<thead>
<tr class="header">
<th><strong>Component/Deliverable</strong></th>
<th><strong>Responsible Team/Service</strong></th>
<th><strong>No earlier than</strong></th>
<th><p><strong>Contact</strong></p>
<p><strong>Person </strong></p></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>HCA Metadata Draft</td>
<td>EBI</td>
<td>Complete</td>
<td>Z. Perova</td>
</tr>
<tr class="even">
<td>Test Datasets Available</td>
<td><em>starfish</em></td>
<td>Complete</td>
<td>A. Carr</td>
</tr>
<tr class="odd">
<td>Submission Datasets Available</td>
<td><em>starfish</em></td>
<td>Q2 2019</td>
<td>A. Carr</td>
</tr>
<tr class="even">
<td>Analysis Pipelines</td>
<td>Broad</td>
<td>Q2 2019</td>
<td>?</td>
</tr>
<tr class="odd">

> <td>Analysis &amp; Adapter Pipelines</td>

Reviewer 11: What's the difference between this line and the above?

<td>Broad</td>
<td>Q2 2019</td>
<td>A. Carr</td>
</tr>
<tr class="even">
<td>Lira subscription</td>
<td>Broad</td>
<td>TBD</td>
<td>S. Ehsan</td>
</tr>
<tr class="odd">
<td>Data Analysis Service <em>starfish</em> Pipeline Tests</td>
<td>Broad</td>
<td>Q2 2019</td>

> <td>R. Wang</td>

Reviewer 11: I think this should probably be G. Grant

In general, I am concerned that the pipeline related deliverables say no earlier than Q2 while the input validation that we had said we needed to be able to run this pipeline says Q3. We will definitely want the ability to do user-specified parameter validation, and perhaps the previously discussed ability to do a quick check on a single tile before processing the entire bundle, before _delivering_ the pipeline. We can of course start work on it sooner and do testing etc.

Author: Updated `s/wang/grant` which means the higher level issue might get hidden. I'll add this to a check list in the description.

</tr>
<tr class="even">

> <td>Specification for input format</td>

Reviewer 11: Does this include appropriate ranges for user submitted parameters that we can use as guardrails or does that fall into the input validation?

Reviewer 2: What would you prefer? In my mind it was in Input validation but I don't have strong preferences.

Reviewer 11: I'm sorry I missed this comment forever ago. I would prefer input validation!
